src/compiler/types/generate-app-types.ts

```diff
@@ -92,29 +92,31 @@ const generateComponentTypesFile = (config: d.Config, buildCtx: d.BuildCtx, areT
   c.push(COMPONENTS_DTS_HEADER);
   c.push(`import { HTMLStencilElement, JSXBase } from "@stencil/core/internal";`);
 
-  // write the import statements for our type declaration file
-  c.push(
-    ...Object.keys(typeImportData).map((filePath) => {
-      const typeData = typeImportData[filePath];
-      let importFilePath: string;
-      if (isAbsolute(filePath)) {
-        importFilePath = normalizePath('./' + relative(config.srcDir, filePath)).replace(/\.(tsx|ts)$/, '');
-      } else {
-        importFilePath = filePath;
-      }
-
-      return `import { ${typeData
-        .sort(sortImportNames)
-        .map((td) => {
-          if (td.localName === td.importName) {
-            return `${td.importName}`;
-          } else {
-            return `${td.localName} as ${td.importName}`;
-          }
-        })
-        .join(`, `)} } from "${importFilePath}";`;
-    })
-  );
+  // Map event type metadata to partial expressions (omitting import/export keywords)
+  // e.g. { TestEvent } from '../path/to/event/test-event.interface';
+  const expressions = Object.keys(typeImportData).map((filePath) => {
+    const typeData = typeImportData[filePath];
+    let importFilePath: string;
+    if (isAbsolute(filePath)) {
+      importFilePath = normalizePath('./' + relative(config.srcDir, filePath)).replace(/\.(tsx|ts)$/, '');
+    } else {
+      importFilePath = filePath;
+    }
+
+    return `{ ${typeData
+      .sort(sortImportNames)
+      .map((td) => {
+        if (td.localName === td.importName) {
+          return `${td.importName}`;
+        } else {
+          return `${td.localName} as ${td.importName}`;
+        }
+      })
+      .join(`, `)} } from "${importFilePath}";`;
+  });
+
+  // Write all import and export statements for event types
+  c.push(...expressions.map((ref) => `import ${ref}`), ...expressions.map((ref) => `export ${ref}`));
```

Comment:
With this change, we don't export only custom event types. We also end up exporting types that are declared on props as well. Take for instance a component spun up using `npm init stencil component` then modified as such:
```diff
+ export type Foo = string;
+ export type Bar = string;

@Component({
  tag: 'my-component',
  styleUrl: 'my-component.css',
  shadow: true,
})
export class MyComponent {
  /**
   * The first name
   */
-  @Prop() first: string;
+  @Prop() first: Foo;

  /**
   * The middle name
   */
-  @Prop() middle: string;
+  @Prop() middle: Bar;
```

running `npm run build` with this branch installed in the project results in the following (note the exports added):
```diff

diff --git a/src/components.d.ts b/src/components.d.ts
index cf9e773..f93c8fc 100644
--- a/src/components.d.ts
+++ b/src/components.d.ts
@@ -5,12 +5,14 @@
  * It contains typing information for all components that exist in this project.
  */
 import { HTMLStencilElement, JSXBase } from "@stencil/core/internal";
+import { Bar, Foo } from "./components/my-component/my-component";
+export { Bar, Foo } from "./components/my-component/my-component";
 export namespace Components {
     interface MyComponent {
         /**
           * The first name
          */
-        "first": string;
+        "first": Foo;
         /**
           * The last name
          */
@@ -18,7 +20,7 @@ export namespace Components {
         /**
           * The middle name
          */
-        "middle": string;
+        "middle": Bar;
     }
 }```

is this intentional?

Reply:
I saw this while working on it as well. I was a little on the fence, but thinking about I don't think there is any risk to these types being exported. This could be helpful for users as well if their properties are expecting complex types they can use these types/interfaces to ensure their type safety without needing to duplicate a type/interface in their consuming app.